client/globals.d.ts

```diff
@@ -186,6 +187,10 @@ declare global {
 		siteTitle: string;
 	};
 
+	const wcSettingsModule: {
+		getSetting: ( setting: string, fallback?: unknown ) => any;
```

Comment:
I set `any` as return here because the `getSetting` function can return pretty much anything. `unkwown` is not really possible as this is now tied to the `WCPayExpressCheckoutParams` interface. Not sure if there would be a better TS approach in this scenario, but sticking with `any` for now.

Reply:
> [!NOTE]
> **TL;DR:** There's a better way to do this, but I'm not sure how it will interact with the rest of the codebase. Worth trying, but I wouldn't spend too much time on it if it's causing problems.

There is, in fact, a better way to do this, but I don't know if it's appropriate to use here. You can look at [this TS playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgcwKYwMrpsAdsuAXzgDMoIQ4ByKVAQwGMYqBuAKDbxlShMdTgBBEBACuueAjZw4DUVFq4GATwBc1UQGcAJlTgAfagzq72MuiPEx1uUSABGPdoQ5cefBgIDiAGwj26HywYHHxNRGk4CzEJdWEYmDNZAAtUBgBrMRgAIVEQiFwAFVQAD2s4TRgoPGQkiDAcAsCAdWAYZIBVXG1UEjxUbQB+dUrq-DqG4CafVvaAOVEfHxGqmoM4WyXnDgB6HbgAKS14HpFcUbpQgnbgcLpw4BI4AHdUGgFQSFg1sgoKilQz1StDgqB8mlQADo2Hs4AAlehMOC3KIANzowB8dHsPgE9yicDAjHSdDQyNwcBu4TAWOUyHI4m0ABp-nAAFbHZEgUn9SnJFEokTaR7KNYAA2eDGCV00YuhPQYWJBQsWAnejGYERkDAKlRQ6GlawAvHAADwAaVBZVQ3XC6VQyggT18-kChrCLMKVu4trgAAoXQEgtgapoANrmgC660ZvX62gAlAA+P32tRwc0svhLAIZYZwQoJ9SBt0hsIR6OGQrbGH7I31uDNAQ6xbaKLgiBwbSdqkvNrJOBiKCD54Uz64kA2mCXKYUp1wMVoTBl5Bill4F78hgD4wQl4CDl65J0VECbvI+A6hRpGA+ZSQuD1k17Xb7OYQHCePmXOAl4MhUM4G5ZQNg-OBHC7ONcAGaFYR1c54GQPwg3dZBNGLZDSwAsIY26KCBkfOBYz6aDtHYWtZF1RCDRXQiLW9G1tDtB15z-VDNE9BjfQDTD-xlCtcJ6EiBmTVMHXUTNSECHxc3SfNCww10+NDASq0fJMIlhGRHn9FBePYwZITDNNowTTT9hkGRaBgeQKSQpT2OMh1I3YLSiFfSydL9KSc2JOAzKkNyrPQWyfJkvz8WrCiZBcaK4FhJsNlQAj2gEMVIrFSllDAPFNAhb4CjgYEBBgTt7VQMA+QEazQpgbLm2gWgmFgiy4BqqAKWI+MonCKLYVi2EnzgABRbpB3kLkaVQScJBnQrSgsKaHyGl8PPfMLZLgdEfFEVBwknOhzgXLrSMylFDsHSYCmheC9WiKxCKXVC-Soe6JCoBN2Fuy9UgyLJcnyIpSngE0npXF7tzSTI8gB0qgbKD6vqoy7GlwQIuiEx6aOw5AXvqVGWn7DH8N0T62G+lHZ0CBYlix5ccbxq60ZmfsaZ8RG4A8gAxaTNu23bwjRhQIGeKqspyyke1SSjryRIkoAsdAeC2wIBZauB30-ARkhFqqQToEFcE7E6CP5vaip4VByeRt6YFmZIed8jI6ee17LHelkkDka8lHTKhjF0FlbfUABmAAGCOiDJinIb+mG8jh4oyntx3wud0HsauCHfuhnIE4KJPmBZKgAAU6BAo3nkRtaPxK494DFTYfDO8JKkxHwCW24A20YK9hXwSWxbqiXNGUSppvAtI6C0NLIXxqmWfmRZm9kC6INKGlgAYNo7wqdBB8b5e5Qowp+QeO4mFEaSQJ1U9aDbDdUsmicpzmilDrbYxcCoeBjzAHLcAsnsHkOAABJIqJ4bQ-1IMAZA8gCJZCKrrUqQE6D2iHvVUe49KC0BILiJE7RLjq1ARSDwMAWSEPgCAWByR4BG3gBBUQlUUFP2HgILB3BKCOGMDPC8VBwhTmALQXeXc2woPahSQ+SxMrFXJOPEwg4nhP2zGndIKsdpW1hDpSh-DNw-ifuOaar9UYvEOjAcIKDuw3WRvPaYbMU68z8hnemWcqC2OZvbNmVBi70IALRN2rrCDAisoj-x8MAc2KCxQm20DIy2PUFxN0ytiCAp5rEIUptMYmwltAOKdmo5xrt3GE3aNk+M3jqB+JidXCiAARCA5t6EvGgOkB8XQdQgBmvAFBEI64CB4OQKAmh1YU3bqgZAgQXbgyoN2Pa-iPy+NKLcZgZM4LIzGRMjuhTpm2wqfw1Y+B5kwF8b3Cq3BSauX2KMpY4zJnbIZv7HO-187wyLogQc9g2Q3nQtQA2AgmmnIaAMPQhBVlXPWTczZUyHnFMXskLxLIQ5kyAA) I set up for some inspiration on the precise implementation. I'll try to give an overview here and would defer to you on which is better; using `any` or narrowing the function signature.

So if we want to narrow the type here we need 3 things:

1. Define which `setting` string is associated with what type of object;
2. Define the type of object returned for each `setting` string; and then
3. Narrow the function signature to reflect that.

Based on what I see in the PR I think that would look something like this:

```ts
interface WcSettings {
  // I made this optional since it probably won't be loaded every time.
  ece_data?: WCPayExpressCheckoutParams;

  // Other things can be added here in the future to expand the type.
}

const wcSettingsModule: {
  getSettings: <K extends keyof WcSettings, T extends WcSettings[K] | undefined>( setting: K, fallback?: T ): WcSettings[K] | T;
};
```

It's best to refer to the TS playground to make sense of the type signature. The short version is that we make sure to only accept known, valid keys to the function, and then use that to infer valid types for the fallback value and possible return values.

This would give us a lot of type safety, but only in the context of what we define ourselves. It's also possible this would conflict with other uses of the function across the codebase, so if doing this causes issues I'd recommend just using `any`. `any` is mostly fine here, even if the narrowed interface would be a good improvement for existing and future code.

It just might require a lot more work to implement. Worth trying though!

Reply:
>we make sure to only accept known, valid keys to the function, and then use that to infer valid types for the fallback value and possible return values

I think the benefit in changing to your suggested approach is what you described here. Changed it in 38da8e0.